src/ripple/nodestore/ShardValidation.md

```diff
@@ -0,0 +1,130 @@
+
+```C++
+boost::optional<uint256>
+hashOfSeq (ReadView const& ledger,
+           LedgerIndex seq,
+           beast::Journal journal)
+```
+
+When validating downloaded shards, we use this function to retrieve the hash of the shard's last ledger. If the function returns a hash that matches the hash stored in the shard, validation of the shard can proceed.
+
+### Caveats
+
+#### Later Ledger
+
+The `getHashBySeq` function will provide the hash of a flag ledger only if the application has stored a later ledger. When validating a downloaded shard, if there is no later ledger stored, validation of the shard will be deferred until a later ledger has been stored.
+
+We employ a simple heuristic for determining whether the application has stored a ledger later than the last ledger of the downloaded shard:
+
+```C++
+// We use the presence (or absense) of the validated
+// ledger as a heuristic for determining whether or
+// not we have stored a ledger that comes after the
+// last ledger in this shard. A later ledger must
+// be present in order to reliably retrieve the hash
+// of the shard's last ledger.
+if (app_.getLedgerMaster().getValidatedLedger())
+{
+    auto const hash = app_.getLedgerMaster().getHashBySeq(
+        lastLedgerSeq(shardIndex));
+
+    .
+    .
+    .
+}
+```
+
+The `getHashBySeq` function will be invoked only when a call to `LedgerMaster::getValidatedLedger` returns a validated ledger, rather than a `nullptr`. Otherwise validation of the shard will be deferred.
+
+### Retries
+
+#### Retry Limit
+
+If the server must defer shard validation, the software will initiate a timer that upon expiration, will re-attempt confirming the last ledger hash. We place an upper limit on the number of attempts the server makes to achieve this confirmation. When the maximum number of attempts has been reached, validation of the shard will fail, resulting in the removal of the shard. An attempt counts toward the limit only when we are able to get a validated ledger (implying a current view of the network), but are unable to retrieve the last ledger hash. Retries that occur because no validated ledger was available are not counted.
+
+#### ShardInfo
+
+The `DatabaseShardImp` class stores a container of `ShardInfo` structs, each of which contains information pertaining to a shard held by the server. These structs will be used during shard import to store the last ledger hash (when available) and to track the number of hash confirmation attempts that have been made.
+
+```C++
+struct ShardInfo
+{
+    .
+    .
+    .
+
+    // Used to limit the number of times we attempt
+    // to retrieve a shard's last ledger hash, when
+    // the hash should have been found. See
+    // scheduleFinalizeShard(). Once this limit has
+    // been exceeded, the shard has failed and is
+    // removed.
```

Comment:
The shard will be in the shard store, but its `ShardInfo::State` will be `import` rather than `final` to indicate that its not validated. These states were there before I joined so maybe @miguelportilla can chime in to confirm that the shards are considered invalid by the code unless the state is set to `final`.

Reply:
N/A per the redesign.